pkg/middleware/allowed_hosts.go

```diff
@@ -0,0 +1,116 @@
+/****************************************************************************
+ * Copyright 2020, Optimizely, Inc. and contributors                        *
+ *                                                                          *
+ * Licensed under the Apache License, Version 2.0 (the "License");          *
+ * you may not use this file except in compliance with the License.         *
+ * You may obtain a copy of the License at                                  *
+ *                                                                          *
+ *    http://www.apache.org/licenses/LICENSE-2.0                            *
+ *                                                                          *
+ * Unless required by applicable law or agreed to in writing, software      *
+ * distributed under the License is distributed on an "AS IS" BASIS,        *
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. *
+ * See the License for the specific language governing permissions and      *
+ * limitations under the License.                                           *
+ ***************************************************************************/
+
+// Package middleware //
+package middleware
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"strings"
+)
+
+// AllowedHosts returns a middleware function that rejects requests whose host value does not match any host in allowedHosts.
+// Request host is determined in the following priority order:
+// 1. X-Forwarded-Host header value
+// 2. Forwarded header host= directive value
+// 3. Host property of request (see Host under https://golang.org/pkg/net/http/#Request)
+func AllowedHosts(allowedHosts []string, allowedPort string) func(next http.Handler) http.Handler {
+	allowedMap := make(map[string]bool)
+	for _, allowedHost := range allowedHosts {
+		allowedMap[fmt.Sprintf("%v:%v", allowedHost, allowedPort)] = true
```

Comment:
Is the port always included/required?

Reply:
For determining the request host, it comes from X-Forwarded-Host header, Forwarded header, Host header, or the host portion of the URL. In all those places, port is optional. But, port is required in Agent configuration.

So we do have a problem here: if the user runs Agent on port 80 without https, or 443 with https, then they send requests omitting the port, this middleware will reject those. To solve I can add some additional handling here for those special cases.